File: core/java/src/main/java/hara/lang/protocol/IWorkExecutor.java
package hara.lang.protocol;

import hara.lang.declaration.HaraAvailability;
import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

/** Executes one leaf request from the Work algebra. */
@HaraProtocolBinding(
    namespace = "std.protocol.iworkexecutor",
    name = "IWorkExecutor",
    availability = HaraAvailability.CAPABILITY_GATED,
    capability = "native-runtime-protocols")
public interface IWorkExecutor {
  @HaraMethod(value = "work-execute", arity = 2)
  Object workExecute(Object request);
}

File: core/rust/src/lang/protocol/iworkexecutor.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(
    namespace = "std.protocol.iworkexecutor",
    name = "IWorkExecutor",
    availability = "capability-gated",
    capability = "native-runtime-protocols"
)]
pub trait IWorkExecutor {
    type Request;
    type Output;

    #[hara_method(value = "work-execute", arity = 2)]
    fn work_execute(&self, request: Self::Request) -> Self::Output;
}
